src/oc_erchef/schema/deploy/create_and_update_users.sql

```diff
@@ -0,0 +1,101 @@
+      INSERT INTO keys
+        (id, key_name, public_key, key_version, created_at,  expires_at)
+      VALUES
+         (p_id, 'default', p_public_key, p_pubkey_version, now(), 'infinity'::timestamp);
+    END IF;
+    WITH createduser AS
+      (INSERT INTO users
+        (id, authz_id, username, email, public_key, hashed_password, salt, hash_type,
+         last_updated_by, created_at, updated_at, external_authentication_uid,
+         recovery_authentication_enabled, serialized_object, admin,
+         pubkey_version)
+       VALUES (p_id, p_authz_id, p_username, p_email, 'this_is_not_a_key',
+         p_hashed_password, p_salt, p_hash_type, p_last_updated_by, p_created_at, p_updated_at,
+         p_external_authentication_uid, p_recovery_authentication_enabled,
+         p_serialized_object, p_admin, p_pubkey_version) RETURNING 1)
+      SELECT count(*) FROM createduser INTO inserteduser;
+      RETURN inserteduser;
+  END;
+$$ LANGUAGE plpgsql;
+
+DROP FUNCTION IF EXISTS update_user(integer,text,text,text,password_hash_type,text,text,boolean,text,text,character,timestamp without time zone,boolean,character);
+
+CREATE OR REPLACE FUNCTION update_user(p_pubkey_version users.pubkey_version%TYPE,
+                                       p_public_key users.public_key%TYPE,
+                                       p_hashed_password users.hashed_password%TYPE,
+                                       p_salt users.salt%TYPE,
+                                       p_hash_type users.hash_type%TYPE,
+                                       p_serialized_object users.serialized_object%TYPE,
+                                       p_external_authentication_uid users.external_authentication_uid%TYPE,
+                                       p_recovery_authentication_enabled users.recovery_authentication_enabled%TYPE,
+                                       p_email users.email%TYPE,
+                                       p_username users.username%TYPE,
+                                       p_last_updated_by users.last_updated_by%TYPE,
+                                       p_updated_at users.updated_at%TYPE,
+                                       p_admin users.admin%TYPE,
+                                       p_id character(32)) RETURNS integer AS $$
+    DECLARE
+      updateduser integer;
+    BEGIN
+    IF p_public_key IS NULL THEN
+      DELETE FROM keys WHERE id = p_id AND key_name = 'default';
+    END IF;
+    IF p_public_key != 'this_is_not_a_key' THEN
+      UPDATE keys SET public_key = p_public_key,
+                      key_version = p_pubkey_version,
+                      expires_at = 'infinity'::timestamp
+      WHERE id = p_id AND key_name = 'default';
+        INSERT INTO keys (id, key_name, public_key, key_version, created_at, expires_at)
+        SELECT p_id, 'default', p_public_key, p_pubkey_version, now(), 'infinity'::timestamp
+        WHERE NOT EXISTS (SELECT 1 FROM keys WHERE id = p_id AND key_name = 'default');
```

Comment:
I know this is how we are doing "upsert" in the original code, but I think this construction is racy.  My guess is that concurrent user updates are rare enough that we can stay with this though.

Reply:
Ah, I commented on this down below... if it's available on the Postgres version you're using, an `ON CONFLICT` clause would be better for this.

Alternatively, using CTEs for this might be preferable, as you could contain the operations within a single statement.

You're also in PL/pgSQL, so you could also use exception handling, but that's more expensive.

Reply:
Unfortunately, I'm not sure we can drop support for 9.2 yet (we should check with product) which means no `ON CONFLICT`.  I think overall using `CTE`s would also make some of this a bit more readable; Perhaps we can do that as a more easily reviewable follow-up improvement. (Getting a bit worried that there is a lot going on here as this is already code not many know well.)

Reply:
I have added the new syntax as a comment in there - a TODO for the future!